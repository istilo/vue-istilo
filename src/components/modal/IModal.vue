<script setup lang="ts">
import { reactive, computed, onMounted, useSlots, ref, watch } from 'vue';
import IButton from '@/components/button/IButton.vue';
import IIcon from '@/components/icon/IIcon.vue';

const modal = ref();

const emit = defineEmits(['close', 'backDropClose', 'escapeClose']);
const slots = useSlots();

const state = reactive({
	openClass: '',
	displayClass: '',
	positionClass: '',
});

const props = withDefaults(
	defineProps<{
		/**
		 * Enable the background event close when user clicks
		 */
		hasBackDropClose?: boolean;

		/**
		 * Enable and displayt the close button
		 */
		hasCloseButton?: boolean;

		/**
		 * Enable or disable the close event when the `esc` key pressed
		 */
		hasEscapeClose?: boolean;

		/**
		 * Toggles the modal
		 */
		isOpen: boolean;

		/**
		 * Vertical align the modal
		 */
		isCenter?: boolean;

		/**
		 * Specify the size of the modal
		 * @values container, full, small
		 */
		size?: string;
	}>(),
	{
		hasBackDropClose: true,
		hasCloseButton: true,
		hasEscapeClose: true,
		isOpen: false,
		isCenter: false,
		size: undefined,
	}
);

watch(
	() => props.isOpen,
	(newValue, oldValue) => {
		if (newValue === oldValue) {
			return;
		}

		if (newValue) {
			openModal();
		} else {
			closeModal();
		}

		toggleScrollbar(newValue);
		toggleModal(newValue);
	}
);

onMounted(() => {
	if (props.isOpen) {
		openModal();

		toggleScrollbar(true);
		toggleModal(true);
	}
});

const getModalClasses = computed(() => {
	return [
		'i-modal',
		props.size ? `i-modal-${props.size}` : '',
		state.positionClass,
		state.openClass,
		state.displayClass,
	];
});

const getModalDialogClasses = computed(() => {
	return ['i-modal-dialog', props.isCenter ? 'i-margin-auto-vertical' : ''];
});

const close = () => {
	emit('close');
};

const backDropEvent = (event: { target: any }) => {
	if (modal.value === event.target && props.size !== 'full') {
		emit('backDropClose');
	}
};

const escapeEvent = (event: { keyCode: number }) => {
	if (event.keyCode === 27) {
		emit('escapeClose');
	}
};

const toggleModal = (bool: boolean) => {
	if (bool) {
		state.displayClass = props.isCenter ? 'i-flex' : 'i-display-block';
		state.positionClass = props.isCenter ? 'i-flex-top' : '';

		setTimeout(() => {
			state.openClass = 'i-open';
		}, 100);

		return;
	}

	state.openClass = '';

	setTimeout(() => {
		state.displayClass = '';
	}, 300);
};

const toggleScrollbar = (bool: boolean) => {
	const el = document.body;

	if (bool) {
		el.classList.add('i-modal-page');
		return;
	}

	el.classList.remove('i-modal-page');
};

const openModal = () => {
	if (props.hasEscapeClose) {
		window.addEventListener('keyup', escapeEvent);
	}

	if (props.hasBackDropClose) {
		modal.value.addEventListener('click', backDropEvent);
	}
};

const closeModal = () => {
	if (props.hasEscapeClose) {
		window.removeEventListener('keyup', escapeEvent);
	}

	if (props.hasBackDropClose) {
		modal.value.removeEventListener('click', backDropEvent);
	}
};
</script>

<template>
	<div
		:class="getModalClasses"
		ref="modal">
		<div :class="getModalDialogClasses">
			<div
				v-if="hasCloseButton"
				class="i-close i-modal-close"
				@click="close">
				<i-button
					:isText="true"
					:label="'Close modal'"
					class="i-modal-close-button">
					<i-icon :type="'close'" />
				</i-button>

				<p class="i-modal-close-esc">ESC</p>
			</div>

			<div
				v-if="slots.header"
				class="i-modal-header">
				<slot name="header" />
			</div>

			<div class="i-modal-body">
				<slot />
			</div>

			<div
				v-if="slots.footer"
				class="i-modal-footer">
				<slot name="footer" />
			</div>
		</div>
	</div>
</template>
