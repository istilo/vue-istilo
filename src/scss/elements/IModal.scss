@use 'sass:math';

@import '@/scss/helpers';
@import '@/scss/functions';
@import '@/scss/mixins';

// ============================================================
// Name:            Modal
// Description:     Component to create modal dialogs
//
// Component:       `#{$prefix}modal`
//
// Sub-objects:     `#{$prefix}modal-page`
//                  `#{$prefix}modal-dialog`
//                  `#{$prefix}modal-header`
//                  `#{$prefix}modal-body`
//                  `#{$prefix}modal-footer`
//                  `#{$prefix}modal-title`
//                  `#{$prefix}modal-close`
//
// Adopted:         `#{$prefix}modal-close-default`
//                  `#{$prefix}modal-close-outside`
//                  `#{$prefix}modal-close-full`
//
// Modifiers:       `#{$prefix}modal-container`
//                  `#{$prefix}modal-full`
//
// States:          `#{$prefix}open`
//
// ============================================================

// Variables
// ============================================================

$modal-z-index: layer(modal) !default;
$modal-background: rgba(0, 0, 0, 0.6) !default;
$modal-solid-background: color(secondary) !default;

$modal-padding-horizontal: gutter(xsmall) !default;
$modal-padding-horizontal-s: gutter(small) !default;
$modal-padding-horizontal-m: gutter(medium) !default;
$modal-padding-vertical: $modal-padding-horizontal !default;
$modal-padding-vertical-s: 50px !default;

$modal-dialog-width: 600px !default;
$modal-dialog-background: background-color(inverse) !default;
$modal-dialog-border-radius: border(radius) !default;

$modal-container-width: 1200px !default;

$modal-body-padding-horizontal: gutter(small) !default;
$modal-body-padding-vertical: gutter(small) !default;

$modal-header-padding-horizontal: gutter(small) !default;
$modal-header-padding-vertical: math.div($modal-header-padding-horizontal, 2) !default;
$modal-header-background: color(inverse) !default;

$modal-footer-padding-horizontal: gutter(small) !default;
$modal-footer-padding-vertical: math.div($modal-footer-padding-horizontal, 2) !default;
$modal-footer-background: color(inverse) !default;

$modal-footer-border: border(width) !default;
$modal-footer-border-color: border(color) !default;

$modal-title-font-size: font-size(large) !default;
$modal-title-line-height: 1.5 !default;

$modal-close-position: margin(small) !default;
$modal-close-padding: 5px !default;

$modal-close-outside-position: 0 !default;
$modal-close-outside-translate: 100% !default;
$modal-close-outside-color: darken(color(inverse), 20%) !default;
$modal-close-outside-hover-color: color(inverse) !default;

/* ============================================================
    Component: Modal
   ============================================================ */

/*
 * 1. Hide by default
 * 2. Set position
 * 3. Allow scrolling for the modal dialog
 * 4. Horizontal padding
 * 5. Mask the background-page
 * 6. Fade-in transition
 */

.#{$prefix}modal {
	/* 1 */
	display: none;

	/* 2 */
	position: fixed;
	top: 0;
	bottom: 0;
	left: 0;
	right: 0;
	z-index: $modal-z-index;

	/* 3 */
	overflow-y: auto;
	-webkit-overflow-scrolling: touch;

	/* 4 */
	padding: $modal-padding-vertical $modal-padding-horizontal;

	/* 5 */
	background: $modal-background;

	/* 6 */
	opacity: 0;
	transition: opacity 0.15s linear !important;
}

// Search modal modification
.#{$prefix}modal.#{$prefix}search-modal {
	background: color(inverse);
}

/* Phone landscape and bigger */
@include respond-to(small) {
	.#{$prefix}modal {
		padding: $modal-padding-vertical-s $modal-padding-horizontal-s;
	}
}

/* Tablet landscape and bigger */
@include respond-to(medium) {
	.#{$prefix}modal {
		padding-left: $modal-padding-horizontal-m;
		padding-right: $modal-padding-horizontal-m;
	}
}

/*
 * Open
 */

.#{$prefix}modal.#{$prefix}open {
	opacity: 1;
}

/* Page
   ============================================================ */

/*
 * Prevent scrollbars
 */

.#{$prefix}modal-page {
	overflow: hidden;
}

/* Dialog
   ============================================================ */

/*
 * 1. Create position contex for spinner and close button
 * 2. Dimensions
 * 3. Fix `max-width: 100%` not working in combination with flex and responsive
 *    images in IE11
 *    `!important` needed to overwrite `#{$prefix}width-auto`. See `#modal-medi
 *    -image` in tests
 * 4. Style
 * 5. Slide-in transition
 */

.#{$prefix}modal-dialog {
	/* 1 */
	position: relative;

	/* 2 */
	box-sizing: border-box;
	margin: 0 auto;
	width: $modal-dialog-width;

	/* 3 */
	max-width: unquote('calc(100% - 0.01px)') !important;

	/* 4 */
	background: $modal-dialog-background;
	border-radius: $modal-dialog-border-radius;
	overflow: hidden;

	/* 5 */
	opacity: 0;
	transform: translateY(-100px);
	transition: 0.3s linear;
	transition-property: opacity, transform;
}

.#{$prefix}modal.#{$prefix}search-modal .#{$prefix}modal-dialog {
	background: transparent;
}

/*
 * Open
 */

.#{$prefix}open > .#{$prefix}modal-dialog {
	opacity: 1;
	transform: translateY(0);
}

/* Size modifier
   ============================================================ */

/*
 * Container size
 * Take the same size as the Container component
 */

.#{$prefix}modal-container .#{$prefix}modal-dialog {
	width: $modal-container-width;
}

/*
 * Full size
 * 1. Remove the padding and background from modal
 * 2. Reset all default declartion from modal dialog
 */

/* 1 */
.#{$prefix}modal-full {
	padding: 0;
	background: none;
	border-radius: initial;
}

/* 2 */
.#{$prefix}modal-full .#{$prefix}modal-dialog {
	margin: 0;
	width: 100%;
	max-width: 100%;
	transform: translateY(0);
}

/* Sections
   ============================================================ */

.#{$prefix}modal-body {
	padding: $modal-body-padding-vertical $modal-body-padding-horizontal;
}

.#{$prefix}modal-header {
	padding: $modal-header-padding-vertical $modal-header-padding-horizontal;
	background: $modal-header-background;
}

.#{$prefix}modal-footer {
	border-top: $modal-footer-border solid $modal-footer-border-color;
	padding: $modal-footer-padding-vertical $modal-footer-padding-horizontal;
	background: $modal-footer-background;
}

/*
 * Micro clearfix
 */

.#{$prefix}modal-body,
.#{$prefix}modal-header,
.#{$prefix}modal-footer {
	@include clearfix;
}

/*
 * Remove margin from the last child
 */

.#{$prefix}modal-body > :last-child,
.#{$prefix}modal-header > :last-child,
.#{$prefix}modal-footer > :last-child {
	margin-bottom: 0;
}

/* Title
   ============================================================ */

.#{$prefix}modal-title {
	font-size: $modal-title-font-size;
	line-height: $modal-title-line-height;
	margin: 0;
}

/* Close
 * Adopts `#{$prefix}close`
   ============================================================ */

.#{$prefix}modal-close {
	position: absolute;
	z-index: $modal-z-index;
	top: $modal-close-position;
	right: $modal-close-position;
	padding: $modal-close-padding;

	text-align: center;
	cursor: pointer;
}

.#{$prefix}modal-close-button {
	display: block;
}

.#{$prefix}modal-close-esc {
	display: block;
	position: relative;
	color: color(muted);
	font-size: font-size(small);

	margin: 0;
}

.#{$prefix}modal-alt {
	font-size: 9px;
	color: color('muted');
	margin-top: -5px;
	margin-bottom: 0px;
}

/*
 * Remove margin from adjacent element
 */

[class*='#{$prefix}modal-close']:first-child + * {
	margin-top: 0;
}

/*
 * Outside
 * 1. Prevent scrollbar on small devices
 */

.#{$prefix}modal-close-outside {
	top: $modal-close-outside-position;

	/* 1 */
	right: (-$modal-close-outside-position);
	transform: translate(0, -($modal-close-outside-translate));
	color: $modal-close-outside-color;
}

.#{$prefix}modal-close-outside:hover {
	color: $modal-close-outside-hover-color;
}

@include respond-to(medium) {
	/* 1 */
	.#{$prefix}modal-close-outside {
		right: $modal-close-outside-position;
		transform: translate($modal-close-outside-translate, -($modal-close-outside-translate));
	}
}
