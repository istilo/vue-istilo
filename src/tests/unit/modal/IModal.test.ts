import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { mount, VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import IModal from '@/components/modal/IModal.vue';

enableAutoUnmount(afterEach);

describe('Istilo Modal', () => {
	let wrapper: VueWrapper<any>;

	beforeEach(() => {
		wrapper = mount(IModal, {
			propsData: {
				isOpen: false,
			},
			slots: {
				default: '<p id="modalbody">Modal</p>',
			},
		});

		expect(wrapper.exists()).toBe(true);
	});

	afterEach(() => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Should set default element and class', () => {
		expect(wrapper.classes()).contain('i-modal');
		expect(wrapper.get('.i-modal-body')).toBeDefined();
		expect(wrapper.get('#modalbody')).toBeDefined();
	});

	it('Should set open class', async () => {
		await wrapper.setProps({
			isOpen: true,
		});

		expect(wrapper.classes('i-display-block')).toBe(true);
	});

	it('Should not render the close button', async () => {
		await wrapper.setProps({
			hasCloseButton: false,
		});

		expect(wrapper.find('.i-modal-close').exists()).toBe(false);
	});

	it('Should have center class modifier', async () => {
		await wrapper.setProps({
			isOpen: true,
			isCenter: true,
		});

		expect(wrapper.find('.i-margin-auto-vertical').exists()).toBe(true);
		expect(wrapper.find('.i-flex-top').exists()).toBe(true);
		expect(wrapper.find('.i-flex-top').exists()).toBe(true);
	});

	it('Should have size class modifier', async () => {
		await wrapper.setProps({
			size: 'full',
		});

		expect(wrapper.classes('i-modal-full')).toBe(true);
	});

	it('Should render header and footer slots', () => {
		wrapper = mount(IModal, {
			propsData: {
				isOpen: false,
			},
			slots: {
				default: '<p id="modalbody">Modal</p>',
				header: '<p>header</p>',
				footer: '<p>footer</p>',
			},
		});

		expect(wrapper.find('.i-modal-header').exists()).toBe(true);
		expect(wrapper.find('.i-modal-footer').exists()).toBe(true);
	});
});
