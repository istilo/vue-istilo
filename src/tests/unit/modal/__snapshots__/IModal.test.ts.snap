// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Istilo Modal > Should have center class modifier 1`] = `
"<div class=\\"i-modal i-flex-top i-flex\\">
  <div class=\\"i-modal-dialog i-margin-auto-vertical\\">
    <div class=\\"i-close i-modal-close\\"><button class=\\"i-button i-button-text i-modal-close-button\\" type=\\"button\\" aria-label=\\"Close modal\\" title=\\"Close modal\\"><i class=\\"i-icon i-icon-close\\"></i></button>
      <p class=\\"i-modal-close-esc\\">ESC</p>
    </div>
    <!--v-if-->
    <div class=\\"i-modal-body\\">
      <p id=\\"modalbody\\">Modal</p>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`Istilo Modal > Should have size class modifier 1`] = `
"<div class=\\"i-modal i-modal-full\\">
  <div class=\\"i-modal-dialog\\">
    <div class=\\"i-close i-modal-close\\"><button class=\\"i-button i-button-text i-modal-close-button\\" type=\\"button\\" aria-label=\\"Close modal\\" title=\\"Close modal\\"><i class=\\"i-icon i-icon-close\\"></i></button>
      <p class=\\"i-modal-close-esc\\">ESC</p>
    </div>
    <!--v-if-->
    <div class=\\"i-modal-body\\">
      <p id=\\"modalbody\\">Modal</p>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`Istilo Modal > Should not render the close button 1`] = `
"<div class=\\"i-modal\\">
  <div class=\\"i-modal-dialog\\">
    <!--v-if-->
    <!--v-if-->
    <div class=\\"i-modal-body\\">
      <p id=\\"modalbody\\">Modal</p>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`Istilo Modal > Should render header and footer slots 1`] = `
"<div class=\\"i-modal\\">
  <div class=\\"i-modal-dialog\\">
    <div class=\\"i-close i-modal-close\\"><button class=\\"i-button i-button-text i-modal-close-button\\" type=\\"button\\" aria-label=\\"Close modal\\" title=\\"Close modal\\"><i class=\\"i-icon i-icon-close\\"></i></button>
      <p class=\\"i-modal-close-esc\\">ESC</p>
    </div>
    <div class=\\"i-modal-header\\">
      <p>header</p>
    </div>
    <div class=\\"i-modal-body\\">
      <p id=\\"modalbody\\">Modal</p>
    </div>
    <div class=\\"i-modal-footer\\">
      <p>footer</p>
    </div>
  </div>
</div>"
`;

exports[`Istilo Modal > Should set default element and class 1`] = `
"<div class=\\"i-modal\\">
  <div class=\\"i-modal-dialog\\">
    <div class=\\"i-close i-modal-close\\"><button class=\\"i-button i-button-text i-modal-close-button\\" type=\\"button\\" aria-label=\\"Close modal\\" title=\\"Close modal\\"><i class=\\"i-icon i-icon-close\\"></i></button>
      <p class=\\"i-modal-close-esc\\">ESC</p>
    </div>
    <!--v-if-->
    <div class=\\"i-modal-body\\">
      <p id=\\"modalbody\\">Modal</p>
    </div>
    <!--v-if-->
  </div>
</div>"
`;

exports[`Istilo Modal > Should set open class 1`] = `
"<div class=\\"i-modal i-display-block\\">
  <div class=\\"i-modal-dialog\\">
    <div class=\\"i-close i-modal-close\\"><button class=\\"i-button i-button-text i-modal-close-button\\" type=\\"button\\" aria-label=\\"Close modal\\" title=\\"Close modal\\"><i class=\\"i-icon i-icon-close\\"></i></button>
      <p class=\\"i-modal-close-esc\\">ESC</p>
    </div>
    <!--v-if-->
    <div class=\\"i-modal-body\\">
      <p id=\\"modalbody\\">Modal</p>
    </div>
    <!--v-if-->
  </div>
</div>"
`;
